core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -546,40 +552,41 @@ private[spark] class ExecutorAllocationManager(
 
     override def onStageCompleted(stageCompleted: SparkListenerStageCompleted): Unit = {
       val stageId = stageCompleted.stageInfo.stageId
+      val stageAttemptId = stageCompleted.stageInfo.attemptNumber()
+      val stageAttempt = StageAttempt(stageId, stageAttemptId)
       allocationManager.synchronized {
-        stageIdToNumTasks -= stageId
-        stageIdToNumRunningTask -= stageId
-        stageIdToNumSpeculativeTasks -= stageId
-        stageIdToTaskIndices -= stageId
-        stageIdToSpeculativeTaskIndices -= stageId
-        stageIdToExecutorPlacementHints -= stageId
+        stageAttemptToNumTasks -= stageAttempt
```

Comment:
I think its worth adding a comment here along the lines of "do NOT remove stageAttempt from stageAttemptToNumRunningTasks, because the attempt may still have running tasks, even after another attempt for the stage is submitted."